### ISSUE TITLE
journal: bad `SystemKeepFree` behavior on compressing file systems

### ISSUE THREAD

**Issue by ensc** [systemd/systemd#13411]
When having the journal on a compressing filesystem like UBIFS, the file system can get filled completelyalthough limits have been configured.

E.g. when `/var/log` is on a 100 MB sized UBIFS and `journald.conf` contains

```
SystemMaxUse=3G
SystemKeepFree=20M
```

the file system can get filled  to 100%.  

This happens, because  journald calculates at startup e.g. 400 MB as used (`journalctl --disk-usage` output) and takes this as the upper free limit (instead of the configured 100MB - 20MB = 80 MB ones). 

See [BTRFS](https://btrfs.wiki.kernel.org/index.php/Compression#Why_does_not_du_report_the_compressed_size.3F) documentation for reasons why `st_blocks` on compressing filesystems returns the uncompressed size.

**systemd version the issue has been seen with**

239

**Used distribution**

OpenEmbedded "thud"

**Expected behaviour you didn't see**

`/var/log` will be filled to 80% but never more (`df` output).

**Unexpected behaviour you saw**

`/var/log` will be filled to 100% and syslog becomes dysfunctional.

**Steps to reproduce the problem**

1. place `/var/log` on a small UBIFS partition (e.g. 100 MB). Can be probably reproduced with BTRFS too.

2. configure limits so that 20% stay free

   ```
   SystemKeepFree=20M
   SystemMaxUse=3G
    ```

    (The `SystemMaxUse` seems to be required for the limit mechanism).

3. start to generate log output

4. watch `df /var/log` output

**More details**

- there is running a

  ```conf
  [Service]
  Type=simple
  ExecStart=/bin/bash -c "i=0; while usleep 200000; do echo CAN write error: Resource temporarily unavailable, $i $RANDOM; let ++i; done"
   ```

  job to simulate typical logging

- some reports (where `SystemKeepFree=20M` constraint has been violated):

  ```
  $ df /var/log
  ubi0:log           92780  72188     15820  83% /var/log

  $ journalctl --disk-usage
  Archived and active journals take up 352.1M in the file system.

  $ du -sbh /var/log/
  353M    /var/log/

  $ stat /var/log/journal/537631928e88efaf5d29417fb87bc9f7/system.journal
  File: /var/log/journal/537631928e88efaf5d29417fb87bc9f7/system.journal
  Size: 8388608         Blocks: 16392      IO Block: 4096   regular file
  ```

  ```
  $ journalctl -b -u systemd-journald
  Aug 29 13:51:18 Journal started
  Aug 29 13:51:18 System journal (/var/log/journal/537631928e88efaf5d29417fb87bc9f7) is 352.1M, max 352.1M, 0B free.
  ```

- some hours later (without touching journald):

  ```
  $ df /var/log
  ubi0:log           92780 80704      7304  92% /var/log

  $ journalctl --disk-usage
  Archived and active journals take up 392.1M in the file system.
  ```

  ```
  $ systemctl restart systemd-journald
  $ journalctl -b -u systemd-journald
  Aug 29 13:51:18 System journal (/var/log/journal/537631928e88efaf5d29417fb87bc9f7) is 352.1M, max 352.1M, 0B free.
  Aug 29 16:35:53 Journal stopped
  Aug 29 16:35:53 Journal started
  Aug 29 16:35:53 System journal (/var/log/journal/537631928e88efaf5d29417fb87bc9f7) is 392.1M, max 392.1M, 0B free.
  ```

- and now, the out-of-disk case:

  ```
  $ df /var/log
  ubi0:log           92780 92776         0 100% /var/log

  $ journalctl --disk-usage
  Archived and active journals take up 528.2M in the file system.
  ```

**Comment by poettering** [comment]
So, what do you expect us to do? If file systems give us useless values, things break. No surprise there...

The linked btrfs documentation doesn't really give a good explanation why, it's just a statement how this is implemented, not why.

With a misleading stat() implementation and no other API to do this better there's nothing we can do about this.

Please talk to the kernel fs guys, to fix their APIs for us, then we can do something about this, otherwise I am sorry, but there's nothing we can do.

**Comment by ensc** [comment]
I do not think that there is any chance to change the kernel API because this might programs expecting the existing behavior.

Fix must be done in journald; e.g. it could remove old files when rotating files until the `SystemKeepFree` constraint is satisfied.

**Comment by poettering** [comment]
that won't work. stat()/statfs() data is cached heavily and updated lazily by file systems like btrfs, so they are generally out-of-date. They are not suitable for synchronous algorithms.

**Comment by ensc** [comment]
The caching does not really matter because  there are not required absolutely correct values. When I set a safe `SystemKeepFree` (e.g something like 4x `SystemMaxFileSize`) the `EONSPC` will be prevented, because there is a high timespan between rotation/removal.

**Comment by poettering** [comment]
hmm? you suggested a loop of rotating until constraint is satisfied. that's not going to work. if we remove files until the constraint is satisfied we might delete way too much if statfs()/stat() is lying...

**Comment by ensc** [comment]
When is `statfs()` lying so that this causes a real problem?

IMO ensuring the `SystemKeepFree` constraint by removing files is a much better behavior than the actual one where logging just stops.
